### PR TITLE
fix(search-algolia): pass custom transformItems function to SearchBar

### LINKED ISSUE #8461: Pass transformItems to SearchBar
### Have you read the Contributing Guidelines on issues?

- [X] I have read the [Contributing Guidelines on issues](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#reporting-new-issues).

### Prerequisites

- [X] I'm using the latest version of Docusaurus.
- [X] I have tried the `npm run clear` or `yarn clear` command.
- [X] I have tried `rm -rf node_modules yarn.lock package-lock.json` and re-installing packages.
- [X] I have tried creating a repro with https://new.docusaurus.io.
- [ ] I have read the console error message carefully (if applicable).

### Description

One of Algolia config options should be [transformItems](https://autocomplete-experimental.netlify.app/docs/docsearchmodal/#transformitems) function which can filter/reorder the incoming search results prior to displaying to the user.

However, it seems like this function is not being picked up when defined in Docusaurus config at all.

### Reproducible demo

https://stackblitz.com/edit/github-mfcf4x?file=docusaurus.config.js

### Steps to reproduce

1. Add algolia config to `docusaurus.config.js`:
```
themeConfig: {
  algolia: {
    appId: 'X1Z85QJPUV',
    apiKey: 'bf7211c161e8205da2f933a02534105a',
    indexName: 'docusaurus-2',
    contextualSearch: true,
    transformItems: (items) => {
      console.log(items); // This should log search hits, but does not.
      return items;
    }
  },
}
```
2. Try to search for something that returns results.
3. Observe the console if the search hits are logged.

### Expected behavior

Search hits should be logged to the console.

### Actual behavior

Search hits are not logged to the console.

### Your environment

- Public source code:https://github.com/h2oai/wave/tree/master/website
- Public site URL: https://wave.h2o.ai/
- Docusaurus version used: 2.1.0, but reproducible on latest as well
- Environment name and version (e.g. Chrome 89, Node.js 16.4): Chrome Version 108.0.5359.124 (Official Build) (x86_64)
- Operating system and version (e.g. Ubuntu 20.04.2 LTS): MacOS 10.15.7

### Self-service

- [ ] I'd be willing to fix this bug myself.

### PR DESCRIPTION
**Pull request by mturoci**
## Pre-flight checklist

- [x] I have read the [Contributing Guidelines on pull requests](https://github.com/facebook/docusaurus/blob/main/CONTRIBUTING.md#pull-requests).
- [ ] **If this is a code change**: I have written unit tests and/or added dogfooding pages to fully verify the new behavior.
- [x] **If this is a new API or substantial change**: the PR has an accompanying issue (closes #8461) and the maintainers have approved on my working plan.

## Motivation

Closes #8462

## Test Plan

Manual

### Test links

Deploy preview: https://deploy-preview-_____--docusaurus-2.netlify.app/

## Related issues/PRs

Closes #8461

### CODE REVIEW

File: packages/docusaurus-theme-search-algolia/src/theme/SearchBar/index.tsx [hunk 1]
```diff
@@ -171,21 +171,23 @@ function DocSearch({
...
+              return item;
+            }
+
+            // We transform the absolute URL into a relative URL.
+            const url = new URL(item.url);
+            return {
+              ...item,
+              url: withBaseUrl(`${url.pathname}${url.hash}`),
+            };
+          }),
```
Comment (Josh-Cena) [inline packages/docusaurus-theme-search-algolia/src/theme/SearchBar/index.tsx:190]: Extract the arrow to a new variable and write `useRef(props.transformItems ?? defaultTransformItems)`.

**Comment by slorber** [comment]
This PR references itself (instead of an external issue) and does not really give much context into what you are trying to achieve (although I have an idea, I'd like you to explain)

---

This is also doing to conflict with https://github.com/facebook/docusaurus/pull/8428, that may already solve your use-case?

**Comment by slorber** [comment]
Found the original issue 🤪 https://github.com/facebook/docusaurus/issues/8461

Using a wrapper to provide a custom transformItems prop is exactly what I had in mind, so feel free to go ahead.

Still wonder if the newly introduced code does not solve your use-case? cf https://github.com/facebook/docusaurus/pull/8428 

```
  const transformItems = useRef<DocSearchModalProps['transformItems']>(
    (items) =>
      items.map((item) => ({
        ...item,
        url: processSearchResultUrl(item.url),
      })),
  ).current;
```

Why do you want to transform those items?

**Comment by mturoci** [comment]
Sorry @slorber @Josh-Cena. I was on a Christmas holiday but seems like you guys took care of this PR already.

> This PR references itself

Oops, my bad.

> Still wonder if the newly introduced code does not solve your use-case

Not really. I wanted to tweak the search results and reorder them the way I want. The linked issue allows only changing the results pathname if I could understand correctly.

**Pull request merged**